File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        DummyNode = ListNode(-1)

        DummyNode.next = head

        length = 0
        curr = head
        while curr:
            length += 1
            curr = curr.next

        length  = length - n
        print(length)

        prev, curr = DummyNode, head,

        while curr:
            if length == 0:
                prev.next = curr.next

            prev, curr = curr, curr.next
            length -= 1

        return DummyNode.next
            

        

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        /**
           One fast pontier move 2 times 
           one slow pointe remove 1 time             
        **/
        if(head == null) return false;

        ListNode curr = head.next;
        ListNode prev = null;

        while(curr != null && curr.next != null){
            if(curr == prev)
                return true;
            curr = curr.next.next;
            if(prev == null){
                prev = head;
            }else{
                prev = prev.next;
            }
        }

        return false;
    }
}
